### PR TITLE
Add API middleware

### PR DESCRIPTION
**Pull request by KostyaTookitaki**

### CODE REVIEW

File: src/actions/__tests__/login.spec.js [hunk 1]
```diff
@@ -17,15 +15,6 @@ describe('Test for Login actions', () => {
     expect(actual).toEqual(expected);
   });
 
-  it('create LOGOUT_USER action', () => {
```
Comment (KostyaTookitaki) [inline src/actions/__tests__/login.spec.js:20]: Logout user action changed to use api middleware

File: src/actions/login.js [hunk 2]
```diff
@@ -4,10 +4,22 @@ export const LOGIN_USER = 'LOGIN_USER';
...
+export const logoutUser = () => ({
+  type: LOGOUT_USER,
+  meta: {
+    api: true
+  },
+  payload: {
+    path: 'users/session',
+    method: 'delete',
+    onSuccess: [clearUserData],
+    onError: [clearUserData]
```
Comment (KostyaTookitaki) [inline src/actions/login.js:23]: `onSuccess` and `onError` can be array of action creators, but can be single, here used array just to show that it supports it

File: src/components/AppLayout.jsx [hunk 3]
```diff
@@ -84,12 +84,13 @@ const SiteFooter = styled(Footer)`
...
     bottom: 0;
     width: 100vw;
+    height: 70px;
   }
 `;
 
 const SiteContent = styled(Content)`
   &&& {
-    margin: 24px 16px 0;
+    margin: 24px 16px 70px;
```
Comment (KostyaTookitaki) [inline src/components/AppLayout.jsx:93]: footer overlaps content so we need margin here

File: src/components/Box.jsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  ${color}
+  ${space}
+  ${typography}
+  ${flexbox}
+  ${background}
+  ${border}
+  ${position}
+`;
+
+export default Box;
```
Comment (KostyaTookitaki) [inline src/components/Box.jsx:24]: this is very useful component which we can use in order to construct layout using flexbox and stuff...

File: src/components/Dashboard.jsx [hunk 5]
```diff
@@ -1,24 +1,69 @@
...
+    dispatch(fetchProducts());
+  }, [dispatch]);
+
+  useRequestIsDone(FETCH_PRODUCTS, () => {
+    if (Boolean(productsError)) {
+      console.log('Products loaded with the error:', productsError);
+    } else {
+      console.log('Products loaded without an error');
+    }
+  });
```
Comment (KostyaTookitaki) [inline src/components/Dashboard.jsx:38]: added it just to have an example that we can trigger custom logic when request is finished

File: src/components/Loading.jsx [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import { Spin } from 'antd';
 import styled from 'styled-components/macro';
 
-const Loading = () => {
-  return <StyledSpin size="large" />;
+const Loading = ({ className }) => {
+  return <StyledSpin className={className} size="large" />;
```
Comment (KostyaTookitaki) [inline src/components/Loading.jsx:5]: in order to allow further styling of it by styled-components we need to support `className` property, same goes for any reusable component

File: src/hooks/general.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+export function usePrevious(value) {
+  const ref = useRef(null);
+
+  useEffect(() => {
+    ref.current = value;
+  }, [value]);
+
+  return ref.current;
+}
```
Comment (KostyaTookitaki) [inline src/hooks/general.js:11]: this is useful hook from official react docs

File: src/middlewares/api.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    } = payload;
+
+    next(action);
+    dispatch(clearNetworkError(actionType));
+    dispatch(startNetwork(actionType));
+
+    try {
+      const headers = {
+        ...actionHeaders,
+        Authorization: `Token ${getLoginDetails(getState()).token}`
```
Comment (KostyaTookitaki) [inline src/middlewares/api.js:41]: adding auth token by defautl

File: src/middlewares/api.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      }
+
+      dispatch(endNetwork(actionType));
+    } catch (error) {
+      console.error('API error', error, action);
+      dispatch(setNetworkError(actionType, error));
+
+      if (get(error, 'response.status') === 401) {
+        dispatch(logoutUser());
+        history.push(PATH.defaultPath);
```
Comment (KostyaTookitaki) [inline src/middlewares/api.js:62]: log user out on `401` response

File: src/middlewares/api.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+import {
+  startNetwork,
+  endNetwork,
+  clearNetworkError,
+  setNetworkError
+} from '../actions/network';
+import { logoutUser } from '../actions/login';
+
+export function apiMiddleware({ getState, dispatch }) {
```
Comment (KostyaTookitaki) [inline src/middlewares/api.js:17]: in order to have better understanding better to look at Redux developer tools browser extension and see which action are dispatched when API request is being performed

File: src/actions/products.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+export const FETCH_PRODUCTS = 'FETCH_PRODUCTS';
+export const SET_PRODUCTS = 'SET_PRODUCTS';
+
+export const setProducts = actionCreator(SET_PRODUCTS, 'payload');
+
+export const fetchProducts = () => ({
+  type: FETCH_PRODUCTS,
+  meta: {
+    api: true
```
Comment (KostyaTookitaki) [inline src/actions/products.js:11]: only action with such meta will go through API middleware

**Review by anishkr07 — APPROVED** [review]

**Pull request merged**